[fx][subgraph_matcher] Add a matcher that supports name to node map

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110743

Summary:
We want the matcher to return a name -> node in target graph
so that we can refer to the node by name, this is useful for downstream applications like
quantization.

and also we can use the torch API as source of truth instead of matching aten API directly.

Test Plan:
python test/fx/test_matcher_utils.py

Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
[fx][subgraph_matcher] Add a matcher that supports name to nodes map

Summary:
We want the matcher to return a name -> node in target graph
so that we can refer to the node by name, this is useful for downstream applications like
quantization.

Test Plan:
python test/fx/test_matcher_utils.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[fx][subgraph_matcher] Add a matcher that supports name to nodes map"


Summary:
We want the matcher to return a name -> node in target graph
so that we can refer to the node by name, this is useful for downstream applications like
quantization.

Test Plan:
python test/fx/test_matcher_utils.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[fx][subgraph_matcher] Add a matcher that supports name to nodes map"


Summary:
We want the matcher to return a name -> node in target graph
so that we can refer to the node by name, this is useful for downstream applications like
quantization.

Test Plan:
python test/fx/test_matcher_utils.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

SherlockNoMad commented on torch/fx/passes/utils/matcher_utils.py:
```diff
@@ -380,3 +387,41 @@ def backtracking(anchor_index, match):
+    return gm, name_nodes_map
+
+
+class SubgraphMatcherWithNameNodesMap(SubgraphMatcher):
```
Need documentation on this. Especially 

- Pattern graph's output convention about `{"conv": conv, "relu": relu}`
- match()'s return InternalMatch. name_nodes_map now has a different meaning.

SherlockNoMad commented on torch/fx/passes/utils/matcher_utils.py:
```diff
@@ -2,14 +2,17 @@
 import os
 
-__all__ = ['SubgraphMatcher', 'InternalMatch']
+__all__ = ['SubgraphMatcher', 'InternalMatch', 'SubgraphMatcherWithNameNodesMap']
```
I am not sure how widely useful this will be for OSS user. 

I suggest keeping this "private", or specific to quantization workflow.

jerryzh168 replied:
this is going to be official API for quant workflow though?

SherlockNoMad replied:
yes, I would prefer this in a different file.

SherlockNoMad reviewed (changes requested):
per comment.

jerryzh168 committed:
Update on "[fx][subgraph_matcher] Add a matcher that supports name to node map"


Summary:
We want the matcher to return a name -> node in target graph
so that we can refer to the node by name, this is useful for downstream applications like
quantization.

Test Plan:
python test/fx/test_matcher_utils.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[fx][subgraph_matcher] Add a matcher that supports name to node map"


Summary:
We want the matcher to return a name -> node in target graph
so that we can refer to the node by name, this is useful for downstream applications like
quantization.

Test Plan:
python test/fx/test_matcher_utils.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[fx][subgraph_matcher] Add a matcher that supports name to node map"


Summary:
We want the matcher to return a name -> node in target graph
so that we can refer to the node by name, this is useful for downstream applications like
quantization.

and also we can use the torch API as source of truth instead of matching aten API directly.

Test Plan:
python test/fx/test_matcher_utils.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot merge

jerryzh168 committed:
Update on "[fx][subgraph_matcher] Add a matcher that supports name to node map"


Summary:
We want the matcher to return a name -> node in target graph
so that we can refer to the node by name, this is useful for downstream applications like
quantization.

and also we can use the torch API as source of truth instead of matching aten API directly.

Test Plan:
python test/fx/test_matcher_utils.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[fx][subgraph_matcher] Add a matcher that supports name to node map"


Summary:
We want the matcher to return a name -> node in target graph
so that we can refer to the node by name, this is useful for downstream applications like
quantization.

and also we can use the torch API as source of truth instead of matching aten API directly.

Test Plan:
python test/fx/test_matcher_utils.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[fx][subgraph_matcher] Add a matcher that supports name to node map"


Summary:
We want the matcher to return a name -> node in target graph
so that we can refer to the node by name, this is useful for downstream applications like
quantization.

and also we can use the torch API as source of truth instead of matching aten API directly.

Test Plan:
python test/fx/test_matcher_utils.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[fx][subgraph_matcher] Add a matcher that supports name to node map"


Summary:
We want the matcher to return a name -> node in target graph
so that we can refer to the node by name, this is useful for downstream applications like
quantization.

and also we can use the torch API as source of truth instead of matching aten API directly.

Test Plan:
python test/fx/test_matcher_utils.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot merge
